{
 "cells": [
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 2,
=======
   "execution_count": 217,
>>>>>>> refs/remotes/origin/master
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import zarr\n",
<<<<<<< HEAD
    "import allel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zarr2LDNe\n",
    " \n",
    "       Sanjay C Nagi      06/07/20\n",
    "\n",
    "I have written a snakemake pipeline which subsets and downsamples the zarr genotype arrays in phase 2, and runs LDNe on the populations within. I now want to run this on all of phase 3. As phase 3 is grouped by sample_set, I want to run LDNe on all locations + years that have greater than 15 samples, though this number is currently arbitrary.\n",
    "\n",
    "Which groups of samples do we want to run LDNe on? see below...\n",
    "### Phase 3 samples for LDNe"
=======
    "import allel\n",
    "import pybedtools"
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>species</th>\n",
       "      <th>year</th>\n",
       "      <th>n_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Angola</td>\n",
       "      <td>Luanda</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2009</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2014</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Daiguene</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Gado Badzere</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Mayos</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Zembe Borongo</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2013</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2005</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2013</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Central African Republic</td>\n",
       "      <td>Bangui</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>1994</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Democratic Republic of Congo</td>\n",
       "      <td>Gbadolite</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2015</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Gabon</td>\n",
       "      <td>Libreville</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2000</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Njabakunda</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2011</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Wali Kunda</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Wali Kunda</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Koforidua</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>Koraboh</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>Koundara</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Antula</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2010</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Antula</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2010</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Safim</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2010</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Kenya</td>\n",
       "      <td>Kilifi</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2000</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Kenya</td>\n",
       "      <td>Kilifi</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Malawi</td>\n",
       "      <td>Chikhwawa</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2015</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Kababougou</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Takan</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Toumani Oulena</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Mali</td>\n",
       "      <td>misc</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2004</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Mali</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2004</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Mayotte</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2011</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Furvela</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2004</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Moshi</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muheza</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2013</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muleba</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2015</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muleba</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2015</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Tarime</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         country        location  \\\n",
       "0                         Angola          Luanda   \n",
       "2                   Burkina Faso            Bana   \n",
       "3                   Burkina Faso            Bana   \n",
       "4                   Burkina Faso            Bana   \n",
       "5                   Burkina Faso            Bana   \n",
       "10                  Burkina Faso            Pala   \n",
       "11                  Burkina Faso            Pala   \n",
       "12                  Burkina Faso  Souroukoudinga   \n",
       "14                  Burkina Faso  Souroukoudinga   \n",
       "15                  Burkina Faso  Souroukoudinga   \n",
       "16                      Cameroon        Daiguene   \n",
       "17                      Cameroon    Gado Badzere   \n",
       "18                      Cameroon           Mayos   \n",
       "19                      Cameroon   Zembe Borongo   \n",
       "22                      Cameroon            misc   \n",
       "23                      Cameroon            misc   \n",
       "24                      Cameroon            misc   \n",
       "28      Central African Republic          Bangui   \n",
       "29                 Cote d'Ivoire        Tiassale   \n",
       "30  Democratic Republic of Congo       Gbadolite   \n",
       "32                         Gabon      Libreville   \n",
       "35                   Gambia, The      Njabakunda   \n",
       "42                   Gambia, The      Wali Kunda   \n",
       "44                   Gambia, The      Wali Kunda   \n",
       "46                         Ghana       Koforidua   \n",
       "49                         Ghana        Takoradi   \n",
       "50                         Ghana     Twifo Praso   \n",
       "51                        Guinea         Koraboh   \n",
       "54                        Guinea        Koundara   \n",
       "55                 Guinea-Bissau          Antula   \n",
       "56                 Guinea-Bissau          Antula   \n",
       "59                 Guinea-Bissau           Safim   \n",
       "62                         Kenya          Kilifi   \n",
       "64                         Kenya          Kilifi   \n",
       "65                        Malawi       Chikhwawa   \n",
       "67                          Mali      Kababougou   \n",
       "70                          Mali           Takan   \n",
       "75                          Mali  Toumani Oulena   \n",
       "78                          Mali            misc   \n",
       "79                          Mali            misc   \n",
       "80                       Mayotte            misc   \n",
       "82                    Mozambique         Furvela   \n",
       "83                      Tanzania           Moshi   \n",
       "85                      Tanzania          Muheza   \n",
       "87                      Tanzania          Muleba   \n",
       "88                      Tanzania          Muleba   \n",
       "90                      Tanzania          Tarime   \n",
       "92                        Uganda          Kihihi   \n",
       "93                        Uganda       Nagongera   \n",
       "94                        Uganda       Nagongera   \n",
       "\n",
       "                          species  year  n_samples  \n",
       "0                        coluzzii  2009         81  \n",
       "2                        coluzzii  2012         42  \n",
       "3                        coluzzii  2014         47  \n",
       "4                         gambiae  2012         22  \n",
       "5                         gambiae  2014         15  \n",
       "10                        gambiae  2012         48  \n",
       "11                        gambiae  2014         16  \n",
       "12                       coluzzii  2012         29  \n",
       "14                        gambiae  2012         28  \n",
       "15                        gambiae  2014         15  \n",
       "16                        gambiae  2009         96  \n",
       "17                        gambiae  2009         73  \n",
       "18                        gambiae  2009        110  \n",
       "19                        gambiae  2009         24  \n",
       "22                       coluzzii  2013         19  \n",
       "23                        gambiae  2005         90  \n",
       "24                        gambiae  2013         23  \n",
       "28                        gambiae  1994         53  \n",
       "29                       coluzzii  2012         80  \n",
       "30                        gambiae  2015         76  \n",
       "32                        gambiae  2000         69  \n",
       "35                        gambiae  2011         58  \n",
       "42                       coluzzii  2012        148  \n",
       "44  intermediate_gambiae_coluzzii  2012         24  \n",
       "46                        gambiae  2012         23  \n",
       "49                       coluzzii  2012         24  \n",
       "50                       coluzzii  2012         25  \n",
       "51                        gambiae  2012         60  \n",
       "54                        gambiae  2012         63  \n",
       "55                        gambiae  2010         15  \n",
       "56  intermediate_gambiae_coluzzii  2010         45  \n",
       "59  intermediate_gambiae_coluzzii  2010         27  \n",
       "62                        gambiae  2000         19  \n",
       "64  intermediate_gambiae_coluzzii  2012         45  \n",
       "65                     arabiensis  2015         41  \n",
       "67                        gambiae  2014         28  \n",
       "70                       coluzzii  2012         26  \n",
       "75                        gambiae  2012         60  \n",
       "78                       coluzzii  2004         36  \n",
       "79                        gambiae  2004         33  \n",
       "80                        gambiae  2011         23  \n",
       "82                        gambiae  2004         71  \n",
       "83                     arabiensis  2012         40  \n",
       "85                        gambiae  2013         36  \n",
       "87                     arabiensis  2015        137  \n",
       "88                        gambiae  2015         32  \n",
       "90                     arabiensis  2012         47  \n",
       "92                        gambiae  2012         95  \n",
       "93                     arabiensis  2012         81  \n",
       "94                        gambiae  2012        112  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest = pd.read_csv(\"../../data/phase3/Ag1000g.phase3.manifest.tsv\", sep=\"\\t\")\n",
    "manifest[manifest.n_samples >= 15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The idea here would be to (in snakemake) loop through each sample_set (not shown above), then within each sample set loop through each location and finally year, where there is temporal samples. \n",
    "\n",
    "At each point we need to extract the appropriate genotypes from the sample set zarrs. We'll also need to filter SNPs, and then can run the downstream LDNe prep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "manifest = pd.read_csv(\"../../data/phase3/Ag1000g.phase3.manifest.full.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>partner_sample_id</th>\n",
       "      <th>contributor</th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>year</th>\n",
       "      <th>month</th>\n",
       "      <th>latitude</th>\n",
       "      <th>longitude</th>\n",
       "      <th>sex_call</th>\n",
       "      <th>sample_set</th>\n",
       "      <th>aim_fraction_colu</th>\n",
       "      <th>aim_fraction_arab</th>\n",
       "      <th>species_gambcolu_arabiensis</th>\n",
       "      <th>species_gambiae_coluzzii</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>AB0085-Cx</td>\n",
       "      <td>BF2-4</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>11.150</td>\n",
       "      <td>-4.235</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>0.024</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>AB0086-Cx</td>\n",
       "      <td>BF2-6</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>11.150</td>\n",
       "      <td>-4.235</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>gambiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>AB0087-C</td>\n",
       "      <td>BF3-3</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>11.233</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>0.982</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>AB0088-C</td>\n",
       "      <td>BF3-5</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>11.233</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>0.990</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>AB0089-Cx</td>\n",
       "      <td>BF3-8</td>\n",
       "      <td>Austin Burt</td>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>2012</td>\n",
       "      <td>7</td>\n",
       "      <td>11.233</td>\n",
       "      <td>-4.472</td>\n",
       "      <td>F</td>\n",
       "      <td>AG1000G-BF-A</td>\n",
       "      <td>0.975</td>\n",
       "      <td>0.002</td>\n",
       "      <td>gamb_colu</td>\n",
       "      <td>coluzzii</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id partner_sample_id  contributor       country location  year  \\\n",
       "81  AB0085-Cx             BF2-4  Austin Burt  Burkina Faso     Pala  2012   \n",
       "82  AB0086-Cx             BF2-6  Austin Burt  Burkina Faso     Pala  2012   \n",
       "83   AB0087-C             BF3-3  Austin Burt  Burkina Faso     Bana  2012   \n",
       "84   AB0088-C             BF3-5  Austin Burt  Burkina Faso     Bana  2012   \n",
       "85  AB0089-Cx             BF3-8  Austin Burt  Burkina Faso     Bana  2012   \n",
       "\n",
       "    month  latitude  longitude sex_call    sample_set  aim_fraction_colu  \\\n",
       "81      7    11.150     -4.235        F  AG1000G-BF-A              0.024   \n",
       "82      7    11.150     -4.235        F  AG1000G-BF-A              0.038   \n",
       "83      7    11.233     -4.472        F  AG1000G-BF-A              0.982   \n",
       "84      7    11.233     -4.472        F  AG1000G-BF-A              0.990   \n",
       "85      7    11.233     -4.472        F  AG1000G-BF-A              0.975   \n",
       "\n",
       "    aim_fraction_arab species_gambcolu_arabiensis species_gambiae_coluzzii  \n",
       "81              0.002                   gamb_colu                  gambiae  \n",
       "82              0.002                   gamb_colu                  gambiae  \n",
       "83              0.002                   gamb_colu                 coluzzii  \n",
       "84              0.002                   gamb_colu                 coluzzii  \n",
       "85              0.002                   gamb_colu                 coluzzii  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest[manifest.sample_set == _set].head()"
=======
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.read_csv(\"~/ag1000g/data/samples.meta.txt\", sep=\"\\t\")"
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
<<<<<<< HEAD
    "### Phase 3"
=======
    "As the current method of running LDNe in the snakemake pipeline requires subsetting the VCF (very slow), I will instead write a script to convert the zarr to .dat format for LDNe."
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 254,
   "metadata": {},
   "outputs": [],
   "source": [
    "_set = 'AG1000G-BF-A'\n",
    "chrom = '3L'\n",
    "n = 10000"
=======
   "execution_count": 236,
   "metadata": {},
   "outputs": [],
   "source": [
    "chrom='3L'\n",
    "\n",
    "Ag_store = zarr.open_array(f\"/home/sanj/ag1000g/data/ag1000g.phase2.ar1.pass/{chrom}/calldata/GT/\", mode = 'r')\n",
    "positions = allel.SortedIndex(zarr.open_array(f\"/home/sanj/ag1000g/data/ag1000g.phase2.ar1.pass/{chrom}/variants/POS\", mode='r'))\n",
    "\n",
    "ag_geno = allel.GenotypeChunkedArray(Ag_store)"
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ag_array = zarr.open_array(f\"/home/sanj/ag1000g/data/phase3/snp_genotypes/all/{_set}/{chrom}/calldata/GT/\", mode = 'r')\n",
    "filters = zarr.open(f\"/home/sanj/ag1000g/data/phase3/site_filters/dt_20200416/gamb_colu/{chrom}/variants/filter_pass\", mode=\"r\")\n",
    "positions = zarr.open_array(f\"/home/sanj/ag1000g/data/phase3/snp_genotypes/all/sites/{chrom}/variants/POS/\", mode='r')\n",
    "metadata = manifest[manifest.sample_set == _set].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtering out coding regions 3L\n",
      "Pala 2012 gambiae (48, 15)\n",
      "Converting to .dat AG1000G-BF-A Pala 2012 gambiae 3L\n",
      "Bana 2012 coluzzii (42, 15)\n",
      "Converting to .dat AG1000G-BF-A Bana 2012 coluzzii 3L\n",
      "Souroukoudinga 2012 coluzzii (29, 15)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-256-358c4f02b001>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     58\u001b[0m             \u001b[0;31m# MAF 0.05 filter\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m             \u001b[0mac\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount_alleles\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 60\u001b[0;31m             \u001b[0mfreqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mac\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_frequencies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     61\u001b[0m             \u001b[0mALT1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfreqs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0.05\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m             \u001b[0mALT2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfreqs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0;36m0.05\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    164\u001b[0m         \u001b[0mdask\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbase\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m         \"\"\"\n\u001b[0;32m--> 166\u001b[0;31m         \u001b[0;34m(\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtraverse\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/site-packages/dask/base.py\u001b[0m in \u001b[0;36mcompute\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    442\u001b[0m         \u001b[0mpostcomputes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__dask_postcompute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    443\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 444\u001b[0;31m     \u001b[0mresults\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mschedule\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdsk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeys\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    445\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mrepack\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ma\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpostcomputes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    446\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/site-packages/dask/threaded.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(dsk, result, cache, num_workers, pool, **kwargs)\u001b[0m\n\u001b[1;32m     74\u001b[0m                 \u001b[0mpools\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mthread\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mnum_workers\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpool\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     75\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 76\u001b[0;31m     results = get_async(\n\u001b[0m\u001b[1;32m     77\u001b[0m         \u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply_async\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     78\u001b[0m         \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpool\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_pool\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/site-packages/dask/local.py\u001b[0m in \u001b[0;36mget_async\u001b[0;34m(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)\u001b[0m\n\u001b[1;32m    473\u001b[0m             \u001b[0;31m# Main loop, wait on tasks to finish, insert new ones\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    474\u001b[0m             \u001b[0;32mwhile\u001b[0m \u001b[0mstate\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"waiting\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mstate\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ready\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mor\u001b[0m \u001b[0mstate\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"running\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 475\u001b[0;31m                 \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mres_info\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfailed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mqueue_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mqueue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    476\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0mfailed\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    477\u001b[0m                     \u001b[0mexc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres_info\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/site-packages/dask/local.py\u001b[0m in \u001b[0;36mqueue_get\u001b[0;34m(q)\u001b[0m\n\u001b[1;32m    131\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    132\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mqueue_get\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 133\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    134\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    135\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/queue.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m    168\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    169\u001b[0m                 \u001b[0;32mwhile\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_qsize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 170\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mnot_empty\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    171\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    172\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"'timeout' must be a non-negative number\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/apps/anaconda3/envs/NeAg/lib/python3.8/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    300\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    301\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 302\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    303\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    304\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "geno = allel.GenotypeDaskArray(Ag_array)\n",
    "geno = geno[filters[:]]\n",
    "positions = positions[:][filters[:]]\n",
    "\n",
    "#filter the gff3 to be coding and regulatory regions\n",
    "df = allel.gff3_to_dataframe(f\"../../data/reference/gff/An.gambiae-PEST-BASEFEATURES_agamP4.12.gff3.gz\")\n",
=======
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter the gff3 to be coding and regulatory regions\n",
    "df = allel.gff3_to_dataframe(\"/home/sanj/ag1000g/data/reference/Anopheles-gambiae-PEST_BASEFEATURES_AgamP4.12.gff3\")\n",
>>>>>>> refs/remotes/origin/master
    "coding_reg_df = df[~df.type.isin(['chromosome', 'three_prime_UTR','five_prime_UTR',\n",
    "                  'mRNA', 'CDS', 'exon'])].drop(columns=['source', 'strand', 'phase', 'score'])\n",
    "coding_reg_df = coding_reg_df[coding_reg_df.seqid == chrom]\n",
    "\n",
<<<<<<< HEAD
    "#get non-centromeric regions. currently chosen by eye based on ag1000g phase1 paper fig1.\n",
    "if chrom == '2L':\n",
    "    centromere = (positions > 3000000)\n",
    "elif chrom == '2R':\n",
    "    centromere = (positions < 57000000)\n",
    "elif chrom == '3L':\n",
    "    centromere = (positions > 2000000)\n",
    "elif chrom == '3R':\n",
    "    centromere = (positions < 50000000)\n",
    "elif chrom == 'X':\n",
    "    centromere = (positions < 21000000)\n",
    "\n",
    "positions = allel.SortedIndex(positions[centromere])\n",
    "#get boolean array for positions that are coding - allel.locate_ranges so fast!\n",
    "coding = positions.locate_ranges(coding_reg_df.start, coding_reg_df.end, strict=False)\n",
    "\n",
    "#compress to get noncoding SNPs and remove centromeric regions of low recombination\n",
    "#TODO currently using all non-coding regions, alternate option may be to take SNPs x distance from coding regions \n",
    "geno = geno.compress(centromere, axis=0)\n",
    "geno = geno.compress(~coding, axis=0) #we want noncoding regions so '~' to get inverse of boolean\n",
    "positions = positions[~coding]\n",
    "print(f\"Filtering out coding regions {chrom}\")\n",
    "\n",
    "\n",
    "### loop through combos \n",
    "for loc in metadata.location.unique():\n",
    "    \n",
    "    nmeta = metadata[metadata.location == loc]\n",
    "    \n",
    "    for yr in nmeta.year.unique():\n",
    "        \n",
    "        nmeta = nmeta[nmeta.year == yr]\n",
    "        \n",
    "        for sp in nmeta.species_gambiae_coluzzii.unique():\n",
    "            \n",
    "            #if there is less than 15 samples than skip\n",
    "            if (nmeta.species_gambiae_coluzzii == sp).sum() < 15:\n",
    "                continue\n",
    "            \n",
    "            #filter to species \n",
    "            nmeta = nmeta[nmeta.species_gambiae_coluzzii == sp]\n",
    "            flt = np.array(nmeta.index)\n",
    "            print(loc, yr, sp, nmeta.shape)\n",
    "            \n",
    "            #filter to correct loc, year, species individuals\n",
    "            gn = geno.take(flt, axis=1)\n",
    "                       \n",
    "            # MAF 0.05 filter\n",
    "            ac = gn.count_alleles()\n",
    "            freqs = ac.to_frequencies().compute()\n",
    "            ALT1 = freqs[:,1] > 0.05\n",
    "            ALT2 = freqs[:,2] > 0.05\n",
    "            ALT3 = freqs[:,3] > 0.05\n",
    "            maf_flt = np.logical_or(ALT1, ALT2, ALT3)\n",
    "            gn = gn.compress(maf_flt, axis=0)\n",
    "\n",
    "            #take random sample of n SNPs \n",
    "            snp_sample = np.random.choice(gn.shape[0], n, replace=False)\n",
    "            snp_sample.sort()\n",
    "            gnr = gn.take(snp_sample, axis=0)\n",
    "            gnr = np.array(gnr[:])\n",
    "            gnr = gnr.astype(str)\n",
    "\n",
    "            pos = positions[maf_flt]\n",
    "            pos = pos[snp_sample]\n",
    "            prefix = f'{chrom}_'\n",
    "            pos_string = [prefix + p for p in pos.astype(str)]\n",
    "\n",
    "            gnr[gnr == '-1'] = '00' #convert missing alleles \n",
    "            dat = np.empty([gnr.shape[0], gnr.shape[1]])\n",
    "\n",
    "            #join genotypes in same individual \n",
    "            print(f\"Converting to .dat {_set} {loc} {yr} {sp} {chrom}\")\n",
    "            for x in range(gnr.shape[0]):\n",
    "                for y in range(gnr.shape[1]):\n",
    "                    dat[x,y] = ''.join(gnr[x,y])\n",
    "\n",
    "            #convert to .dat format genotypes \n",
    "            dat = dat.astype(str)\n",
    "            dat[dat == '0.0'] = '0101'\n",
    "            dat[dat == '1.0'] = '0102'\n",
    "            dat[dat == '2.0'] = '0103'\n",
    "            dat[dat == '3.0'] = '0104'\n",
    "            dat[dat == '10.0'] = '0201'\n",
    "            dat[dat == '11.0'] = '0202'\n",
    "            dat[dat == '12.0'] = '0203'\n",
    "            dat[dat == '13.0'] = '0204'\n",
    "            dat[dat == '20.0'] = '0301'\n",
    "            dat[dat == '21.0'] = '0302'\n",
    "            dat[dat == '22.0'] = '0303'\n",
    "            dat[dat == '23.0'] = '0304'\n",
    "            dat[dat == '30.0'] = '0401'\n",
    "            dat[dat == '31.0'] = '0402'\n",
    "            dat[dat == '32.0'] = '0403'\n",
    "            dat[dat == '33.0'] = '0404'\n",
    "\n",
    "            #stack population name and transposed genotypes \n",
    "            popnames = np.repeat(f\"{_set}{loc}{yr}{sp}{chrom}\", gnr.shape[1])\n",
    "            dat = np.column_stack((popnames, dat.T)) #\n",
    "\n",
    "            #write out .dat file for LDNe \n",
    "            with open(f'../data/dat/{_set}.{loc}.{yr}.{sp}.{chrom}.dat', 'w') as datfile:\n",
    "                datfile.write(f'{gnr.shape[1]}\\t{gnr.shape[0]}\\t4\\t2\\n')\n",
    "                datfile.write(\"\\n\".join(\"\".join(map(str, x)) for x in pos_string)) \n",
    "                datfile.write(\"\\n\")\n",
    "                datfile.write(\"\\n\".join(\"\\t\".join(map(str, x)) for x in dat))    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_sets = manifest.sample_set.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "ag = ag_geno.compress((positions[~coding] > 35000000), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_alleles = ag.count_alleles().sum(axis=1)[:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "ac = ag.count_alleles()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1152356,)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first = ac.to_frequencies()[:][:,1] > 0.05\n",
    "second = ac.to_frequencies()[:][:,2] > 0.05\n",
    "third = ac.to_frequencies()[:][:,3] > 0.05\n",
    "\n",
    "np.logical_or(first, second, third).shape"
=======
    "centromere = (positions > 2000000) if chrom == '3L' else (positions > 50000000)\n",
    "positions = positions[centromere]\n",
    "#get boolean array for positions that are coding\n",
    "coding = positions.locate_ranges(coding_reg_df.start, coding_reg_df.end, strict=False)\n",
    "\n",
    "#compress to get noncoding SNPs\n",
    "ag_geno = ag_geno.compress(centromere, axis=0)\n",
    "ag_geno = ag_geno.compress(~coding, axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove centromeric regions of low recombination\n",
    "\n",
    "Converting to .dat format for LDNe"
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
=======
   "execution_count": 238,
>>>>>>> refs/remotes/origin/master
   "metadata": {},
   "outputs": [],
   "source": [
    "pops = samples.population.unique()"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 7,
=======
   "execution_count": 239,
>>>>>>> refs/remotes/origin/master
   "metadata": {},
   "outputs": [],
   "source": [
    "pops = pops[2:3]"
   ]
  },
  {
   "cell_type": "code",
<<<<<<< HEAD
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['BFgam'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for chrom in chroms:\n",
    "\n",
    "    #open arrays\n",
    "    print(f\"Opening arrays {chrom}\")\n",
    "    Ag_store = zarr.open_array(f\"{args.zarr}/{chrom}/calldata/GT/\", mode = 'r')\n",
    "    positions = allel.SortedIndex(zarr.open_array(f\"{args.zarr}/{chrom}/variants/POS\", mode='r'))\n",
    "    ag_geno = allel.GenotypeChunkedArray(Ag_store)\n",
    "\n",
    "    #filter the gff3 to be coding and regulatory regions\n",
    "    df = allel.gff3_to_dataframe(f\"{args.gff}\")\n",
    "    coding_reg_df = df[~df.type.isin(['chromosome', 'three_prime_UTR','five_prime_UTR',\n",
    "                      'mRNA', 'CDS', 'exon'])].drop(columns=['source', 'strand', 'phase', 'score'])\n",
    "    coding_reg_df = coding_reg_df[coding_reg_df.seqid == chrom]\n",
    "\n",
    "    #get non-centromeric regions. currently chosen by eye based on ag1000g phase1 paper fig1.\n",
    "    if chrom == '2L':\n",
    "        centromere = (positions > 3000000)\n",
    "    elif chrom == '2R':\n",
    "        centromere = (positions < 57000000)\n",
    "    elif chrom == '3L':\n",
    "        centromere = (positions > 2000000)\n",
    "    elif chrom == '3R':\n",
    "        centromere = (positions < 50000000)\n",
    "    elif chrom == 'X':\n",
    "        centromere = (positions < 21000000)\n",
    "\n",
    "    positions = positions[centromere]\n",
    "    #get boolean array for positions that are coding - allel.locate_ranges so fast!\n",
    "    coding = positions.locate_ranges(coding_reg_df.start, coding_reg_df.end, strict=False)\n",
    "\n",
    "    #compress to get noncoding SNPs and remove centromeric regions of low recombination\n",
    "    #TODO currently using all non-coding regions, alternate option may be to take SNPs x distance from coding regions \n",
    "    ag_geno = ag_geno.compress(centromere, axis=0)\n",
    "    ag_geno = ag_geno.compress(~coding, axis=0) #we want noncoding regions so '~' to get inverse of boolean\n",
    "    print(f\"Filtering out coding regions {chrom}\")\n",
    "\n",
    "    # Converting to .dat format for LDNe\n",
    "    for pop in pops:\n",
    "        pop_bool = samples.population == pop\n",
    "        pop_geno = ag_geno.compress(pop_bool, axis=1)\n",
    "\n",
    "        # MAF 0.05 filter\n",
    "        ac = pop_geno.count_alleles()\n",
    "        freqs = ac.to_frequencies()[:]\n",
    "        ALT1 = freqs[:,1] > 0.05\n",
    "        ALT2 = freqs[:,2] > 0.05\n",
    "        ALT3 = freqs[:,3] > 0.05\n",
    "        maf_flt = np.logical_or(ALT1, ALT2, ALT3)\n",
    "\n",
    "        pop_geno = pop_geno.compress(maf_flt, axis=0)\n",
    "\n",
    "        #take random sample of n SNPs \n",
    "        snp_sample = np.random.choice(pop_geno.shape[0], n, replace=False)\n",
    "        snp_sample.sort()\n",
    "        gnr = pop_geno.take(snp_sample, axis=0)\n",
    "        gnr = np.array(gnr[:])\n",
    "        gnr = gnr.astype(str)\n",
    "\n",
    "        pos = positions[~coding]\n",
    "        pos = pos[seg]\n",
    "        pos = pos[vidx]\n",
    "        prefix = f'{chrom}_'\n",
    "        pos_string = [prefix + p for p in pos.astype(str)]\n",
    "\n",
    "        gnr[gnr == '-1'] = '00' #convert missing alleles \n",
    "        dat = np.empty([gnr.shape[0], gnr.shape[1]])\n",
    "\n",
    "        #join genotypes in same individual \n",
    "        print(f\"Converting to .dat {pop} {chrom}\")\n",
    "        for x in range(gnr.shape[0]):\n",
    "            for y in range(gnr.shape[1]):\n",
    "                dat[x,y] = ''.join(gnr[x,y])\n",
    "\n",
    "        #convert to .dat format genotypes \n",
    "        dat = dat.astype(str)\n",
    "        dat[dat == '0.0'] = '0101'\n",
    "        dat[dat == '1.0'] = '0102'\n",
    "        dat[dat == '2.0'] = '0103'\n",
    "        dat[dat == '3.0'] = '0104'\n",
    "        dat[dat == '10.0'] = '0201'\n",
    "        dat[dat == '11.0'] = '0202'\n",
    "        dat[dat == '12.0'] = '0203'\n",
    "        dat[dat == '13.0'] = '0204'\n",
    "        dat[dat == '20.0'] = '0301'\n",
    "        dat[dat == '21.0'] = '0302'\n",
    "        dat[dat == '22.0'] = '0303'\n",
    "        dat[dat == '23.0'] = '0304'\n",
    "        dat[dat == '30.0'] = '0401'\n",
    "        dat[dat == '31.0'] = '0402'\n",
    "        dat[dat == '32.0'] = '0403'\n",
    "        dat[dat == '33.0'] = '0404'\n",
    "\n",
    "        #stack population name and transposed genotypes \n",
    "        popnames = np.repeat(f\"{pop}_{chrom}\", gnr.shape[1])\n",
    "        dat = np.column_stack((popnames, dat.T)) #\n",
    "        \n",
    "        #write out .dat file for LDNe \n",
    "        with open(f'data/dat/{pop}_{chrom}.dat', 'w') as datfile:\n",
    "            datfile.write(f'{gnr.shape[1]}\\t{gnr.shape[0]}\\t4\\t2\\n')\n",
    "            datfile.write(\"\\n\".join(\"\".join(map(str, x)) for x in pos_string)) \n",
    "            datfile.write(\"\\n\")\n",
    "            datfile.write(\"\\n\".join(\"\\t\".join(map(str, x)) for x in dat))    "
=======
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----------- Zarr to LDNe BFgam 3L ------------\n",
      "Compressing array to BFgam 3L\n",
      "Counting alleles\n",
      "Selecting segregating alleles\n",
      "Converting to .dat format\n",
      "Writing to file BFgam 3L\n"
     ]
    }
   ],
   "source": [
    "#pops = samples.population.unique()\n",
    "\n",
    "for pop in pops:\n",
    "    print(f\"----------- Zarr to LDNe {pop} {chrom} ------------\")\n",
    "    pop_bool = samples.population == pop\n",
    "    pop_geno = ag_geno.compress(pop_bool, axis=1)\n",
    "    print(f\"Compressing array to {pop} {chrom}\")\n",
    "    \n",
    "    ac = pop_geno.count_alleles() ; print(\"Counting alleles\")\n",
    "    seg = ac.is_segregating()\n",
    "    pop_geno = pop_geno.compress(seg, axis=0) ; print(\"Selecting segregating alleles\")\n",
    "\n",
    "    n = 10000  # number of SNPs to choose randomly\n",
    "    vidx = np.random.choice(pop_geno.shape[0], n, replace=False)\n",
    "    vidx.sort()\n",
    "    gnr = pop_geno.take(vidx, axis=0)\n",
    "    gnr = np.array(gnr[:])\n",
    "    gnr = gnr.astype(str)\n",
    "\n",
    "    pos = positions[~coding]\n",
    "    pos = pos[seg]\n",
    "    pos = pos[vidx]\n",
    "    prefix = f'{chrom}_'\n",
    "    pos_string = [prefix + p for p in pos.astype(str)]\n",
    "\n",
    "    \n",
    "    print(\"Converting to .dat format\")\n",
    "    dat = np.empty([gnr.shape[0], gnr.shape[1]])\n",
    "    \n",
    "    gnr[gnr == '-1'] = '00' # remove missing alleles\n",
    "    gnr[gnr == '0'] = '01' #convert to dat format allele numbering\n",
    "    gnr[gnr == '1'] = '02'\n",
    "    gnr[gnr == '2'] = '03'\n",
    "    gnr[gnr == '3'] = '04'\n",
    "    \n",
    "    for x in range(gnr.shape[0]):\n",
    "        for y in range(gnr.shape[1]):\n",
    "            dat[x,y] = ''.join(gnr[x,y])\n",
    "\n",
    "    dat = dat.astype(str)\n",
    "\n",
    "    popnames = np.repeat(f\"{pop}_{chrom}\", n)\n",
    "    dat = np.column_stack((popnames, dat))\n",
    "    print(f\"Writing to file {pop} {chrom}\")\n",
    "    with open(f'{pop}_{chrom}.dat', 'w') as datfile:\n",
    "        datfile.write(f'{gnr.shape[1]}\\t{gnr.shape[0]}\\t4\\t2\\n')\n",
    "        datfile.write(\"\\n\".join(\"\".join(map(str, x)) for x in pos_string)) \n",
    "        datfile.write(\"\\n\")\n",
    "        datfile.write(\"\\n\".join(\" \".join(map(str, x)) for x in dat))    "
>>>>>>> refs/remotes/origin/master
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
<<<<<<< HEAD
   "source": []
=======
   "source": [
    " dat[dat == '0.0'] = '0101'\n",
    "    dat[dat == '1.0'] = '0102'\n",
    "    dat[dat == '2.0'] = '0103'\n",
    "    dat[dat == '3.0'] = '0104'\n",
    "    dat[dat == '10.0'] = '0201'\n",
    "    dat[dat == '11.0'] = '0202'\n",
    "    dat[dat == '12.0'] = '0203'\n",
    "    dat[dat == '13.0'] = '0204'\n",
    "    dat[dat == '20.0'] = '0301'\n",
    "    dat[dat == '21.0'] = '0302'\n",
    "    dat[dat == '22.0'] = '0303'\n",
    "    dat[dat == '23.0'] = '0304'\n",
    "    dat[dat == '30.0'] = '0401'\n",
    "    dat[dat == '31.0'] = '0402'\n",
    "    dat[dat == '32.0'] = '0403'\n",
    "    dat[dat == '33.0'] = '0404'"
   ]
>>>>>>> refs/remotes/origin/master
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
<<<<<<< HEAD
   "version": "3.8.1"
=======
   "version": "3.7.3"
>>>>>>> refs/remotes/origin/master
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
