{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd; pd.set_option('display.max_rows', 10000)\n",
    "import allel \n",
    "import zarr\n",
    "from IPython.display import HTML, display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples = pd.read_csv(\"../../data/samples.meta.txt\",  sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimates of recent effective population size (*Ne*) on the Ag1000g data\n",
    "##### Sanjay C Nagi\n",
    "\n",
    "In preparation for the LLINEUP trial data, and with a bit of spare time, I began analyses into effective population size of the Ag1000g populations. I now have a snakemake pipeline to apply LDNe and IDBNe to *Ag* WGS data, which I have done, as well as also calculating Ne from theta=4NeMu. *I have not yet attempted to fully interpret the results*. \n",
    "\n",
    "\n",
    "### LDNe\n",
    "\n",
    "Estimates of effective population size were estimated in NeEstimator v2.1 (Do *et al*., 2014), using the Linkage Disequilibrium method (LDNe) of Waples and Do (2008). This method is widely used (the most widely cited of all methods to estimate *Ne*), and generally performs robustly in comparisons between single-sample estimators (see references for further reading)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "Ne = pd.read_csv(\"Ne_analyses.LDNe\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pop</th>\n",
       "      <th>sample_size</th>\n",
       "      <th>Ne_estimate</th>\n",
       "      <th>Jackknife_CI_lower</th>\n",
       "      <th>Jackknife_CI_upper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>3L</td>\n",
       "      <td>KE</td>\n",
       "      <td>47.9</td>\n",
       "      <td>2.4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>3R</td>\n",
       "      <td>KE</td>\n",
       "      <td>47.9</td>\n",
       "      <td>2.8</td>\n",
       "      <td>2.5</td>\n",
       "      <td>3.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>3L</td>\n",
       "      <td>FRgam</td>\n",
       "      <td>23.9</td>\n",
       "      <td>100.0</td>\n",
       "      <td>74.5</td>\n",
       "      <td>148.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>3R</td>\n",
       "      <td>FRgam</td>\n",
       "      <td>24.0</td>\n",
       "      <td>110.2</td>\n",
       "      <td>87.3</td>\n",
       "      <td>147.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3R</td>\n",
       "      <td>GHgam</td>\n",
       "      <td>12.0</td>\n",
       "      <td>117.7</td>\n",
       "      <td>34.1</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>3R</td>\n",
       "      <td>GNcol</td>\n",
       "      <td>4.0</td>\n",
       "      <td>128.5</td>\n",
       "      <td>18.6</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>3L</td>\n",
       "      <td>AOcol</td>\n",
       "      <td>77.9</td>\n",
       "      <td>247.6</td>\n",
       "      <td>134.5</td>\n",
       "      <td>969.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>3L</td>\n",
       "      <td>GHgam</td>\n",
       "      <td>12.0</td>\n",
       "      <td>279.1</td>\n",
       "      <td>98.9</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>3R</td>\n",
       "      <td>GAgam</td>\n",
       "      <td>69.0</td>\n",
       "      <td>298.1</td>\n",
       "      <td>246.7</td>\n",
       "      <td>374.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>3R</td>\n",
       "      <td>AOcol</td>\n",
       "      <td>78.0</td>\n",
       "      <td>302.8</td>\n",
       "      <td>165.8</td>\n",
       "      <td>1198.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>3L</td>\n",
       "      <td>GAgam</td>\n",
       "      <td>69.0</td>\n",
       "      <td>339.8</td>\n",
       "      <td>300.4</td>\n",
       "      <td>390.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>3L</td>\n",
       "      <td>GNgam</td>\n",
       "      <td>40.0</td>\n",
       "      <td>413.3</td>\n",
       "      <td>126.5</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>3R</td>\n",
       "      <td>GNgam</td>\n",
       "      <td>40.0</td>\n",
       "      <td>453.8</td>\n",
       "      <td>153.0</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3R</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>55.0</td>\n",
       "      <td>680.4</td>\n",
       "      <td>224.3</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>3L</td>\n",
       "      <td>CMgam</td>\n",
       "      <td>296.9</td>\n",
       "      <td>719.7</td>\n",
       "      <td>512.2</td>\n",
       "      <td>1154.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>3R</td>\n",
       "      <td>CMgam</td>\n",
       "      <td>297.0</td>\n",
       "      <td>724.6</td>\n",
       "      <td>531.3</td>\n",
       "      <td>1099.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3L</td>\n",
       "      <td>GHcol</td>\n",
       "      <td>55.0</td>\n",
       "      <td>851.2</td>\n",
       "      <td>338.5</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>3R</td>\n",
       "      <td>GM</td>\n",
       "      <td>65.0</td>\n",
       "      <td>941.7</td>\n",
       "      <td>772.6</td>\n",
       "      <td>1202.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>3L</td>\n",
       "      <td>GM</td>\n",
       "      <td>65.0</td>\n",
       "      <td>969.9</td>\n",
       "      <td>783.3</td>\n",
       "      <td>1269.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3R</td>\n",
       "      <td>BFgam</td>\n",
       "      <td>92.0</td>\n",
       "      <td>1699.6</td>\n",
       "      <td>1040.1</td>\n",
       "      <td>4508.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>3R</td>\n",
       "      <td>UGgam</td>\n",
       "      <td>112.0</td>\n",
       "      <td>2304.8</td>\n",
       "      <td>1798.6</td>\n",
       "      <td>3198.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>3L</td>\n",
       "      <td>BFgam</td>\n",
       "      <td>92.0</td>\n",
       "      <td>2503.3</td>\n",
       "      <td>1317.0</td>\n",
       "      <td>21833.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>3L</td>\n",
       "      <td>UGgam</td>\n",
       "      <td>112.0</td>\n",
       "      <td>2515.7</td>\n",
       "      <td>1960.6</td>\n",
       "      <td>3500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>3L</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>71.0</td>\n",
       "      <td>3203.3</td>\n",
       "      <td>2252.3</td>\n",
       "      <td>5524.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>3R</td>\n",
       "      <td>CIcol</td>\n",
       "      <td>71.0</td>\n",
       "      <td>3204.5</td>\n",
       "      <td>2325.1</td>\n",
       "      <td>5139.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>3L</td>\n",
       "      <td>GW</td>\n",
       "      <td>91.0</td>\n",
       "      <td>4054.2</td>\n",
       "      <td>2338.7</td>\n",
       "      <td>14862.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>3R</td>\n",
       "      <td>GW</td>\n",
       "      <td>91.0</td>\n",
       "      <td>4499.7</td>\n",
       "      <td>3051.0</td>\n",
       "      <td>8529.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>3L</td>\n",
       "      <td>GQgam</td>\n",
       "      <td>9.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>Infinite</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>3R</td>\n",
       "      <td>GQgam</td>\n",
       "      <td>9.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>Infinite</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>3R</td>\n",
       "      <td>BFcol</td>\n",
       "      <td>75.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>9299.5</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3L</td>\n",
       "      <td>BFcol</td>\n",
       "      <td>75.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>10892.7</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>3L</td>\n",
       "      <td>GNcol</td>\n",
       "      <td>4.0</td>\n",
       "      <td>inf</td>\n",
       "      <td>7.4</td>\n",
       "      <td>Infinite</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    pop  sample_size  Ne_estimate Jackknife_CI_lower  \\\n",
       "56     3L     KE         47.9          2.4                2.0   \n",
       "60     3R     KE         47.9          2.8                2.5   \n",
       "72     3L  FRgam         23.9        100.0               74.5   \n",
       "76     3R  FRgam         24.0        110.2               87.3   \n",
       "12     3R  GHgam         12.0        117.7               34.1   \n",
       "116    3R  GNcol          4.0        128.5               18.6   \n",
       "88     3L  AOcol         77.9        247.6              134.5   \n",
       "8      3L  GHgam         12.0        279.1               98.9   \n",
       "100    3R  GAgam         69.0        298.1              246.7   \n",
       "92     3R  AOcol         78.0        302.8              165.8   \n",
       "96     3L  GAgam         69.0        339.8              300.4   \n",
       "104    3L  GNgam         40.0        413.3              126.5   \n",
       "108    3R  GNgam         40.0        453.8              153.0   \n",
       "4      3R  GHcol         55.0        680.4              224.3   \n",
       "64     3L  CMgam        296.9        719.7              512.2   \n",
       "68     3R  CMgam        297.0        724.6              531.3   \n",
       "0      3L  GHcol         55.0        851.2              338.5   \n",
       "44     3R     GM         65.0        941.7              772.6   \n",
       "40     3L     GM         65.0        969.9              783.3   \n",
       "20     3R  BFgam         92.0       1699.6             1040.1   \n",
       "36     3R  UGgam        112.0       2304.8             1798.6   \n",
       "16     3L  BFgam         92.0       2503.3             1317.0   \n",
       "32     3L  UGgam        112.0       2515.7             1960.6   \n",
       "120    3L  CIcol         71.0       3203.3             2252.3   \n",
       "124    3R  CIcol         71.0       3204.5             2325.1   \n",
       "48     3L     GW         91.0       4054.2             2338.7   \n",
       "52     3R     GW         91.0       4499.7             3051.0   \n",
       "80     3L  GQgam          9.0          inf           Infinite   \n",
       "84     3R  GQgam          9.0          inf           Infinite   \n",
       "28     3R  BFcol         75.0          inf             9299.5   \n",
       "24     3L  BFcol         75.0          inf            10892.7   \n",
       "112    3L  GNcol          4.0          inf                7.4   \n",
       "\n",
       "    Jackknife_CI_upper  \n",
       "56                 2.9  \n",
       "60                 3.1  \n",
       "72               148.9  \n",
       "76               147.7  \n",
       "12            Infinite  \n",
       "116           Infinite  \n",
       "88               969.3  \n",
       "8             Infinite  \n",
       "100              374.4  \n",
       "92              1198.5  \n",
       "96               390.3  \n",
       "104           Infinite  \n",
       "108           Infinite  \n",
       "4             Infinite  \n",
       "64              1154.8  \n",
       "68              1099.7  \n",
       "0             Infinite  \n",
       "44              1202.8  \n",
       "40              1269.9  \n",
       "20              4508.3  \n",
       "36              3198.8  \n",
       "16             21833.9  \n",
       "32              3500.0  \n",
       "120             5524.2  \n",
       "124             5139.5  \n",
       "48             14862.7  \n",
       "52              8529.8  \n",
       "80            Infinite  \n",
       "84            Infinite  \n",
       "28            Infinite  \n",
       "24            Infinite  \n",
       "112           Infinite  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Ne['Ne_estimate'] = Ne['Ne_estimate'].replace(\"Infinite\", np.inf).astype(float) #change string to np.inf to allow sorting\n",
    "Ne[Ne['AF'] == 'minAF_0.05'].drop(columns=['AF', 'independent_comparisons', \n",
    "                                           'overall_r^2', 'expected_r^2', \n",
    "                                           'Parametric_CI_lower', 'Parametric_CI_upper']).sort_values(by='Ne_estimate')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Qualitatively, we can see that the largest populations, at least in terms of effective population size, are *coluzzii* from Burkina (BFcol) and Cote D'ivoire (CIcol), as well as *gambiae* from Uganda (UGgam) and Burkina (BFgam).\n",
    "\n",
    "The Guinea-Bissau population (GW) also displays a high *Ne*, which we might expect - overall LD will surely be lower in a hybrid population.\n",
    "\n",
    "Ignoring Bioko and Guinean coluzzii for their small sample size and infinity estimates (9,4), the smallest effective population sizes are found in the Kenyan population (KE), *gambiae* from Mayotte (FRgam), as we would expect, but also the ghaniaian *gambiae* population.\n",
    "\n",
    "In phase 1, it was noted that the Gabon (GAgam) and Angolan (AOgam) populations had reduced diversity, and more extensive LD, indicating smaller Ne than other West Africa populations, in agreement with the LDNe estimates above."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Notes \n",
    "\n",
    "- Values of infinity can result from an underestimate of sampling error - if this occurs, the bias correction can result in negative estimates of Ne. This is observed in Bioko (GQgam) and Guinean coluzzii (GNcol), and possibly is contributing to Kenyas miniscule Ne estimates.\n",
    "- Infinity may also result from too large an effective population size to estimate with the sample size given. For example, we get an infinity value for the point estimate of Burkina Faso coluzzii (BFcol). In this case, the lower CI is informative. <br>\n",
    "- I have excluded the parametric CIs, which seemed far too narrow. Both methods to produce confidence intervals, the parametric and jackknife method, may be suboptimal according to Jones *et al*., (2016), unfortunately, the methods they suggest have not yet been implemented in any usable manner."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### IBDNe "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "IBDNe estimates effective population sizes in the recent past based on the size and abundance of shared segments of IBD in a population (Browning & Browning, 2015). A high abundance of IBD tracts would indicate a small *Ne*, and the length of those IBD tracts would indicate when that small *Ne* occured (as they are whittled down in size by recombination over time).\n",
    "\n",
    "- The full estimates range from 0-300 generations ago, however, the authors suggest that the IBD tracts are informative from 4 to approximately 200 generations ago. \n",
    "- Some of the predicted histories of effective population size are particularly volatile, and the confidence intervals vary in size massively. <br>\n",
    "\n",
    "I have attached the IBDNe plots in the attached pdfs. A couple of populations are missing, due to small sample sizes. The harmonic means are as follows, and roughly seem to agree with the LDNe data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>IBDNe_harmonic_means</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KE</td>\n",
       "      <td>2354.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FRgam</td>\n",
       "      <td>18747.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GAgam</td>\n",
       "      <td>20470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CIcol</td>\n",
       "      <td>25168.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GM</td>\n",
       "      <td>27957.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AOcol</td>\n",
       "      <td>30554.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BFgam</td>\n",
       "      <td>47377.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>UGgam</td>\n",
       "      <td>48750.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GHcol</td>\n",
       "      <td>98537.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GNgam</td>\n",
       "      <td>176859.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GW</td>\n",
       "      <td>331434.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>BFcol</td>\n",
       "      <td>1067389.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Population  IBDNe_harmonic_means\n",
       "0          KE                2354.0\n",
       "1       FRgam               18747.0\n",
       "2       GAgam               20470.0\n",
       "3       CIcol               25168.0\n",
       "4          GM               27957.0\n",
       "5       AOcol               30554.0\n",
       "6       BFgam               47377.0\n",
       "7       UGgam               48750.0\n",
       "8       GHcol               98537.0\n",
       "9       GNgam              176859.0\n",
       "10         GW              331434.0\n",
       "11      BFcol             1067389.0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"ibdne/IBDNe_Harmonic_means.txt\", sep=\"\\t\").round()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### estimating Ne from theta=4Nemu"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As all populations share the same mutation rate, in this case, the Ne estimates are simply a function of diversity (wattersons theta in this case, though I have seen others use pi)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------  Arrays loaded GHcol -----------------\n",
      "Counting alleles GHcol 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GHcol -----------------\n",
      "Counting alleles GHcol 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GHgam -----------------\n",
      "Counting alleles GHgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GHgam -----------------\n",
      "Counting alleles GHgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded BFgam -----------------\n",
      "Counting alleles BFgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded BFgam -----------------\n",
      "Counting alleles BFgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded BFcol -----------------\n",
      "Counting alleles BFcol 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded BFcol -----------------\n",
      "Counting alleles BFcol 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded UGgam -----------------\n",
      "Counting alleles UGgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded UGgam -----------------\n",
      "Counting alleles UGgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GM -----------------\n",
      "Counting alleles GM 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GM -----------------\n",
      "Counting alleles GM 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GW -----------------\n",
      "Counting alleles GW 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GW -----------------\n",
      "Counting alleles GW 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded KE -----------------\n",
      "Counting alleles KE 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded KE -----------------\n",
      "Counting alleles KE 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded CMgam -----------------\n",
      "Counting alleles CMgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded CMgam -----------------\n",
      "Counting alleles CMgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded FRgam -----------------\n",
      "Counting alleles FRgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded FRgam -----------------\n",
      "Counting alleles FRgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GQgam -----------------\n",
      "Counting alleles GQgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GQgam -----------------\n",
      "Counting alleles GQgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded AOcol -----------------\n",
      "Counting alleles AOcol 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded AOcol -----------------\n",
      "Counting alleles AOcol 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GAgam -----------------\n",
      "Counting alleles GAgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GAgam -----------------\n",
      "Counting alleles GAgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GNgam -----------------\n",
      "Counting alleles GNgam 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GNgam -----------------\n",
      "Counting alleles GNgam 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GNcol -----------------\n",
      "Counting alleles GNcol 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded GNcol -----------------\n",
      "Counting alleles GNcol 3R\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded CIcol -----------------\n",
      "Counting alleles CIcol 3L\n",
      "Computing theta\n",
      "done\n",
      "-------------------  Arrays loaded CIcol -----------------\n",
      "Counting alleles CIcol 3R\n",
      "Computing theta\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "pops = samples.population.unique()\n",
    "chroms = ['3L', '3R']\n",
    "\n",
    "mu=3.5e-9\n",
    "\n",
    "Ne = dict()\n",
    "Ne_pi = dict()\n",
    "Ne_Ag = dict()\n",
    "Ne_Ag_pi = dict()\n",
    "        \n",
    "for pop in pops:\n",
    "    for chrom in chroms:\n",
    "        \n",
    "        Ag_array  = zarr.open_array(f\"/home/sanj/ag1000g/data/ag1000g.phase2.ar1.pass/{chrom}/calldata/GT/\")\n",
    "        pos  = zarr.open_array(f\"/home/sanj/ag1000g/data/ag1000g.phase2.ar1.pass/{chrom}/variants/POS\")\n",
    "        geno = allel.GenotypeChunkedArray(Ag_array)\n",
    "        print(f\"-------------------  Arrays loaded {pop} -----------------\")\n",
    "     \n",
    "        pop_bool = samples.population == pop\n",
    "        pop_geno = geno.compress(pop_bool, axis=1)\n",
    "        \n",
    "        print(f\"Counting alleles {pop} {chrom}\")\n",
    "        ac = pop_geno.count_alleles()\n",
    "        print(\"Computing theta\")\n",
    "        theta = allel.watterson_theta(pos, ac)\n",
    "        pi = allel.sequence_diversity(pos, ac)\n",
    "        print('done')\n",
    "        Neff = theta/(4*mu)\n",
    "        Neff_pi = pi/(4*mu)\n",
    "        \n",
    "        Ne[chrom] = Neff\n",
    "        Ne_pi[chrom] = Neff_pi\n",
    "    \n",
    "    Ne_Ag[pop] = dict(Ne)\n",
    "    Ne_Ag_pi[pop] = dict(Ne_pi)\n",
    "\n",
    "Ne_theta = pd.DataFrame.from_dict(Ne_Ag).T\n",
    "Ne_pi = pd.DataFrame.from_dict(Ne_Ag_pi).T\n",
    "Ne_theta.columns = ['3L_theta', '3R_theta']\n",
    "Ne_pi.columns = ['3L_pi', '3R_pi']\n",
    "Ne_Ag = pd.concat([Ne_theta, Ne_pi], axis=1)\n",
    "\n",
    "Ne_Ag.round().to_csv(\"Ne_theta_pi_Ag.csv\", index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3L_theta</th>\n",
       "      <th>3R_theta</th>\n",
       "      <th>3L_pi</th>\n",
       "      <th>3R_pi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GHcol</th>\n",
       "      <td>870489.0</td>\n",
       "      <td>952726.0</td>\n",
       "      <td>522127.0</td>\n",
       "      <td>569754.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GHgam</th>\n",
       "      <td>735458.0</td>\n",
       "      <td>816605.0</td>\n",
       "      <td>538470.0</td>\n",
       "      <td>595453.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFgam</th>\n",
       "      <td>1303007.0</td>\n",
       "      <td>1397081.0</td>\n",
       "      <td>545999.0</td>\n",
       "      <td>600311.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BFcol</th>\n",
       "      <td>1126854.0</td>\n",
       "      <td>1242037.0</td>\n",
       "      <td>529589.0</td>\n",
       "      <td>577963.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UGgam</th>\n",
       "      <td>1230586.0</td>\n",
       "      <td>1321118.0</td>\n",
       "      <td>543128.0</td>\n",
       "      <td>596927.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GM</th>\n",
       "      <td>867399.0</td>\n",
       "      <td>945927.0</td>\n",
       "      <td>550725.0</td>\n",
       "      <td>598330.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GW</th>\n",
       "      <td>1154793.0</td>\n",
       "      <td>1259574.0</td>\n",
       "      <td>558378.0</td>\n",
       "      <td>608367.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KE</th>\n",
       "      <td>220627.0</td>\n",
       "      <td>237080.0</td>\n",
       "      <td>353424.0</td>\n",
       "      <td>372746.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CMgam</th>\n",
       "      <td>1645481.0</td>\n",
       "      <td>1767792.0</td>\n",
       "      <td>546319.0</td>\n",
       "      <td>601318.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FRgam</th>\n",
       "      <td>287884.0</td>\n",
       "      <td>321054.0</td>\n",
       "      <td>360686.0</td>\n",
       "      <td>403933.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GQgam</th>\n",
       "      <td>609771.0</td>\n",
       "      <td>670849.0</td>\n",
       "      <td>521109.0</td>\n",
       "      <td>569602.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AOcol</th>\n",
       "      <td>462409.0</td>\n",
       "      <td>489332.0</td>\n",
       "      <td>432775.0</td>\n",
       "      <td>463546.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAgam</th>\n",
       "      <td>589337.0</td>\n",
       "      <td>664073.0</td>\n",
       "      <td>497899.0</td>\n",
       "      <td>552784.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNgam</th>\n",
       "      <td>1055905.0</td>\n",
       "      <td>1152839.0</td>\n",
       "      <td>546874.0</td>\n",
       "      <td>601463.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNcol</th>\n",
       "      <td>499691.0</td>\n",
       "      <td>553950.0</td>\n",
       "      <td>495624.0</td>\n",
       "      <td>547088.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIcol</th>\n",
       "      <td>877532.0</td>\n",
       "      <td>980350.0</td>\n",
       "      <td>523408.0</td>\n",
       "      <td>572778.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        3L_theta   3R_theta     3L_pi     3R_pi\n",
       "GHcol   870489.0   952726.0  522127.0  569754.0\n",
       "GHgam   735458.0   816605.0  538470.0  595453.0\n",
       "BFgam  1303007.0  1397081.0  545999.0  600311.0\n",
       "BFcol  1126854.0  1242037.0  529589.0  577963.0\n",
       "UGgam  1230586.0  1321118.0  543128.0  596927.0\n",
       "GM      867399.0   945927.0  550725.0  598330.0\n",
       "GW     1154793.0  1259574.0  558378.0  608367.0\n",
       "KE      220627.0   237080.0  353424.0  372746.0\n",
       "CMgam  1645481.0  1767792.0  546319.0  601318.0\n",
       "FRgam   287884.0   321054.0  360686.0  403933.0\n",
       "GQgam   609771.0   670849.0  521109.0  569602.0\n",
       "AOcol   462409.0   489332.0  432775.0  463546.0\n",
       "GAgam   589337.0   664073.0  497899.0  552784.0\n",
       "GNgam  1055905.0  1152839.0  546874.0  601463.0\n",
       "GNcol   499691.0   553950.0  495624.0  547088.0\n",
       "CIcol   877532.0   980350.0  523408.0  572778.0"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(\"Ne_theta_pi_Ag.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions for Martin/Dave/Eric/others\n",
    "\n",
    "Should this work remain as a 2-pager, sent to those who would be interested? \n",
    "\n",
    "Or is it worth producing a standalone paper, perhaps by either.... \n",
    "\n",
    "- Adding analyses into runs of homozygosity (ROH), and comparing ROH profiles between populations. (*I have since realised this was pretty much done in original phase 1 paper, though only discussed for Kenya*).\n",
    "\n",
    "- Writing review on studies of effective population size in Anopheles mosquitoes, and include these analyses as part of that.\n",
    "\n",
    "---------\n",
    "And what is the best way to present this data? ... lots of populations, multiple chromosomes etc. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods\n",
    "The methods need checking - as this was initially just exploratory, I made rapid decisions regarding what might be appropriate.\n",
    "\n",
    "- The analysis has been done on whole populations - however, many populations are from multiple sites in the same country, whereas some are just one site. How will this affect analyses? One might certainly expect it to affect IBDNe estimates.\n",
    "\n",
    "#### LDNe\n",
    "- Chromosomes 3L and 3R were chosen for analysis to avoid major inversions, in agreement with the IBDNe analysis conducted by the Kern lab in Phase 1. \n",
    "- Pericentromeric regions of low recombination were removed - the exact values I used were slightly arbitrary and based on plots in phase 1 paper. \n",
    "- SNPs were restricted to non-coding regions - a better option might be to restrict to x-distance away from coding regions (though I doubt this will modify the results significantly)\n",
    "- 10,000 random SNPs were used for each population, and each population has a different random selection, rather than the same 10,000 SNP positions. Is it more appropriate to use the same 10,000? I dont know.\n",
    "- MAFs of 0, 0.01, 0.02, 0.05 were all tested with LDNe. I have presented 0.05 here, as using the lower MAF thresholds gives a larger Ne estimate but with many more infinity values.\n",
    "\n",
    "--------\n",
    "#### IDBNe\n",
    "- Chromosomes 3L and 3R were chosen as above.\n",
    "- Pericentromeric regions of low recombination were removed as above. \n",
    "- Thats it\n",
    "\n",
    "-------\n",
    "#### Theta=4NeMu\n",
    "- I have used wattersons theta across the whole chromosome accessible regions (not restricted to non-coding regions as above)\n",
    "- mu=3.5e-9 (from Drosophila as in Keightley et al. 2009, (Miles et al., 2017, supplementary text))\n",
    "- Ne=theta/(4*mu)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### References"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
