{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import zarr\n",
    "import allel\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zarr2LDNe\n",
    " \n",
    "       Sanjay C Nagi      06/07/20\n",
    "\n",
    "I have written a snakemake pipeline which subsets and downsamples the zarr genotype arrays in phase 2, and runs LDNe on the populations within. I now want to run this on all of phase 3. As phase 3 is grouped by sample_set, I want to run LDNe on all locations + years that have greater than 15 samples, though this number is currently arbitrary.\n",
    "\n",
    "Which groups of samples do we want to run LDNe on? see below...\n",
    "### Phase 3 samples for LDNe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>country</th>\n",
       "      <th>location</th>\n",
       "      <th>species</th>\n",
       "      <th>year</th>\n",
       "      <th>n_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Angola</td>\n",
       "      <td>Luanda</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2009</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2014</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Bana</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Pala</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Burkina Faso</td>\n",
       "      <td>Souroukoudinga</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Daiguene</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Gado Badzere</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Mayos</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>Zembe Borongo</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2009</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2013</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2005</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Cameroon</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2013</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Central African Republic</td>\n",
       "      <td>Bangui</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>1994</td>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Cote d'Ivoire</td>\n",
       "      <td>Tiassale</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Democratic Republic of Congo</td>\n",
       "      <td>Gbadolite</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2015</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Gabon</td>\n",
       "      <td>Libreville</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2000</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Njabakunda</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2011</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Wali Kunda</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Gambia, The</td>\n",
       "      <td>Wali Kunda</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Koforidua</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Takoradi</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Ghana</td>\n",
       "      <td>Twifo Praso</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>Koraboh</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Guinea</td>\n",
       "      <td>Koundara</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Antula</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2010</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Antula</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2010</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Guinea-Bissau</td>\n",
       "      <td>Safim</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2010</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Kenya</td>\n",
       "      <td>Kilifi</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2000</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>Kenya</td>\n",
       "      <td>Kilifi</td>\n",
       "      <td>intermediate_gambiae_coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>Malawi</td>\n",
       "      <td>Chikhwawa</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2015</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Kababougou</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2014</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Takan</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2012</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>Mali</td>\n",
       "      <td>Toumani Oulena</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Mali</td>\n",
       "      <td>misc</td>\n",
       "      <td>coluzzii</td>\n",
       "      <td>2004</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>Mali</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2004</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>Mayotte</td>\n",
       "      <td>misc</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2011</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>Mozambique</td>\n",
       "      <td>Furvela</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2004</td>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Moshi</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muheza</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2013</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muleba</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2015</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Muleba</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2015</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Tanzania</td>\n",
       "      <td>Tarime</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Kihihi</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>arabiensis</td>\n",
       "      <td>2012</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>Uganda</td>\n",
       "      <td>Nagongera</td>\n",
       "      <td>gambiae</td>\n",
       "      <td>2012</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         country        location  \\\n",
       "0                         Angola          Luanda   \n",
       "2                   Burkina Faso            Bana   \n",
       "3                   Burkina Faso            Bana   \n",
       "4                   Burkina Faso            Bana   \n",
       "5                   Burkina Faso            Bana   \n",
       "10                  Burkina Faso            Pala   \n",
       "11                  Burkina Faso            Pala   \n",
       "12                  Burkina Faso  Souroukoudinga   \n",
       "14                  Burkina Faso  Souroukoudinga   \n",
       "15                  Burkina Faso  Souroukoudinga   \n",
       "16                      Cameroon        Daiguene   \n",
       "17                      Cameroon    Gado Badzere   \n",
       "18                      Cameroon           Mayos   \n",
       "19                      Cameroon   Zembe Borongo   \n",
       "22                      Cameroon            misc   \n",
       "23                      Cameroon            misc   \n",
       "24                      Cameroon            misc   \n",
       "28      Central African Republic          Bangui   \n",
       "29                 Cote d'Ivoire        Tiassale   \n",
       "30  Democratic Republic of Congo       Gbadolite   \n",
       "32                         Gabon      Libreville   \n",
       "35                   Gambia, The      Njabakunda   \n",
       "42                   Gambia, The      Wali Kunda   \n",
       "44                   Gambia, The      Wali Kunda   \n",
       "46                         Ghana       Koforidua   \n",
       "49                         Ghana        Takoradi   \n",
       "50                         Ghana     Twifo Praso   \n",
       "51                        Guinea         Koraboh   \n",
       "54                        Guinea        Koundara   \n",
       "55                 Guinea-Bissau          Antula   \n",
       "56                 Guinea-Bissau          Antula   \n",
       "59                 Guinea-Bissau           Safim   \n",
       "62                         Kenya          Kilifi   \n",
       "64                         Kenya          Kilifi   \n",
       "65                        Malawi       Chikhwawa   \n",
       "67                          Mali      Kababougou   \n",
       "70                          Mali           Takan   \n",
       "75                          Mali  Toumani Oulena   \n",
       "78                          Mali            misc   \n",
       "79                          Mali            misc   \n",
       "80                       Mayotte            misc   \n",
       "82                    Mozambique         Furvela   \n",
       "83                      Tanzania           Moshi   \n",
       "85                      Tanzania          Muheza   \n",
       "87                      Tanzania          Muleba   \n",
       "88                      Tanzania          Muleba   \n",
       "90                      Tanzania          Tarime   \n",
       "92                        Uganda          Kihihi   \n",
       "93                        Uganda       Nagongera   \n",
       "94                        Uganda       Nagongera   \n",
       "\n",
       "                          species  year  n_samples  \n",
       "0                        coluzzii  2009         81  \n",
       "2                        coluzzii  2012         42  \n",
       "3                        coluzzii  2014         47  \n",
       "4                         gambiae  2012         22  \n",
       "5                         gambiae  2014         15  \n",
       "10                        gambiae  2012         48  \n",
       "11                        gambiae  2014         16  \n",
       "12                       coluzzii  2012         29  \n",
       "14                        gambiae  2012         28  \n",
       "15                        gambiae  2014         15  \n",
       "16                        gambiae  2009         96  \n",
       "17                        gambiae  2009         73  \n",
       "18                        gambiae  2009        110  \n",
       "19                        gambiae  2009         24  \n",
       "22                       coluzzii  2013         19  \n",
       "23                        gambiae  2005         90  \n",
       "24                        gambiae  2013         23  \n",
       "28                        gambiae  1994         53  \n",
       "29                       coluzzii  2012         80  \n",
       "30                        gambiae  2015         76  \n",
       "32                        gambiae  2000         69  \n",
       "35                        gambiae  2011         58  \n",
       "42                       coluzzii  2012        148  \n",
       "44  intermediate_gambiae_coluzzii  2012         24  \n",
       "46                        gambiae  2012         23  \n",
       "49                       coluzzii  2012         24  \n",
       "50                       coluzzii  2012         25  \n",
       "51                        gambiae  2012         60  \n",
       "54                        gambiae  2012         63  \n",
       "55                        gambiae  2010         15  \n",
       "56  intermediate_gambiae_coluzzii  2010         45  \n",
       "59  intermediate_gambiae_coluzzii  2010         27  \n",
       "62                        gambiae  2000         19  \n",
       "64  intermediate_gambiae_coluzzii  2012         45  \n",
       "65                     arabiensis  2015         41  \n",
       "67                        gambiae  2014         28  \n",
       "70                       coluzzii  2012         26  \n",
       "75                        gambiae  2012         60  \n",
       "78                       coluzzii  2004         36  \n",
       "79                        gambiae  2004         33  \n",
       "80                        gambiae  2011         23  \n",
       "82                        gambiae  2004         71  \n",
       "83                     arabiensis  2012         40  \n",
       "85                        gambiae  2013         36  \n",
       "87                     arabiensis  2015        137  \n",
       "88                        gambiae  2015         32  \n",
       "90                     arabiensis  2012         47  \n",
       "92                        gambiae  2012         95  \n",
       "93                     arabiensis  2012         81  \n",
       "94                        gambiae  2012        112  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest = pd.read_csv(\"../../data/phase3/Ag1000g.phase3.manifest.tsv\", sep=\"\\t\")\n",
    "manifest[manifest.n_samples >= 15]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The idea here would be to (in snakemake) loop through each sample_set (not shown above), then within each sample set loop through each location and finally year, where there is temporal samples. \n",
    "\n",
    "At each point we need to extract the appropriate genotypes from the sample set zarrs. We'll also need to filter SNPs, and then can run the downstream LDNe prep."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "manifest = pd.read_csv(\"../../data/phase3/Ag1000g.phase3.manifest.full.tsv\", sep=\"\\t\")\n",
    "manifest.location = [loc.replace(\" \", \"\") for loc in manifest.location]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phase 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_sets = manifest.sample_set.unique()\n",
    "\n",
    "chrom = '3L'\n",
    "n = 20000\n",
    "\n",
    "local_path = Path(\"/home/sanj/ag1000g/data/phase3/\").expanduser() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['gamb_colu', 'arabiensis', 'intermediate'], dtype=object)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest.species_gambcolu_arabiensis.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------ LDNe_Ag -------------------------------\n",
      "\n",
      "Storing metadata in ../data/Phase3.LDNe.tsv\n"
     ]
    }
   ],
   "source": [
    "print(f\"------------------------ LDNe_Ag -------------------------------\\n\")\n",
    "print(f\"Storing metadata in ../data/Phase3.LDNe.tsv\")\n",
    "with open(f'../data/Phase3.LDNe.tsv', 'w') as metafile:\n",
    "    metafile.write(\"sample_set\\tlocation\\tyear\\tspecies\\tchromosome\\n\")\n",
    "\n",
    "Path('../data/Phase3.LDNe.list').touch()    \n",
    "  \n",
    "\n",
    "for _set in all_sets:\n",
    "  \n",
    "    #subset metadata\n",
    "    metadata = manifest[manifest.sample_set == _set].reset_index(drop=True)\n",
    "\n",
    "    ### loop through combos \n",
    "    for loc in metadata.location.unique():\n",
    "\n",
    "        nmeta = metadata[metadata.location == loc]\n",
    "\n",
    "        for yr in nmeta.year.unique():\n",
    "\n",
    "            nmeta = nmeta[nmeta.year == yr]\n",
    "            \n",
    "            #have edited .species_gambiae_coluzzii column to contain arabiensis instead of NA \n",
    "            for sp in nmeta.species_gambiae_coluzzii.unique():\n",
    "\n",
    "                #if there is less than 15 samples than skip\n",
    "                if (nmeta.species_gambiae_coluzzii == sp).sum() < 15:\n",
    "                    continue\n",
    "         \n",
    "                #write metadata file for samples that are included\n",
    "                with open('../data/Phase3.LDNe.tsv', 'a') as metafile:\n",
    "                    metafile.write(f'{_set}\\t{loc}\\t{yr}\\t{sp}\\t{chrom}\\n')\n",
    "                \n",
    "                with open('../data/Phase3.LDNe.list', 'a') as listfile:\n",
    "                    listfile.write(f'{_set}.{loc}.{yr}.{sp}\\n')\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------------ LDNe_Ag -------------------------------\n",
      "\n",
      "Storing metadata in ../data/Phase3.LDNe.tsv\n",
      "\n",
      "Producing LDNe input for Luanda, 2009, coluzzii. 81 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-AO Luanda 2009 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Pala, 2012, gambiae. 59 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-A Pala 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Bana, 2012, coluzzii. 65 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-A Bana 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Souroukoudinga, 2012, coluzzii. 57 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-A Souroukoudinga 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Bana, 2014, coluzzii. 63 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-B Bana 2014 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Pala, 2014, gambiae. 18 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-B Pala 2014 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Souroukoudinga, 2014, gambiae. 21 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-BF-B Souroukoudinga 2014 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Gbadolite, 2015, gambiae. 76 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CD Gbadolite 2015 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Tiassale, 2012, coluzzii. 80 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CI Tiassale 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Mayos, 2009, gambiae. 110 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CM-A Mayos 2009 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for GadoBadzere, 2009, gambiae. 73 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CM-A GadoBadzere 2009 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for ZembeBorongo, 2009, gambiae. 24 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CM-A ZembeBorongo 2009 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Daiguene, 2009, gambiae. 96 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-CM-A Daiguene 2009 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Libreville, 2000, gambiae. 69 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GA-A Libreville 2000 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for TwifoPraso, 2012, coluzzii. 25 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GH TwifoPraso 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Takoradi, 2012, coluzzii. 24 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GH Takoradi 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Koforidua, 2012, gambiae. 24 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GH Koforidua 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Njabakunda, 2011, gambiae. 74 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GM-A Njabakunda 2011 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for WaliKunda, 2012, coluzzii. 174 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GM-C WaliKunda 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Koundara, 2012, gambiae. 22 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-A Koundara 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Koraboh, 2012, gambiae. 23 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-A Koraboh 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Koraboh, 2012, gambiae. 39 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-B Koraboh 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Koundara, 2012, gambiae. 52 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-B Koundara 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for ToumaniOulena, 2012, gambiae. 63 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-B ToumaniOulena 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Takan, 2012, coluzzii. 31 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GN-B Takan 2012 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Antula, 2010, gambiae. 60 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GW Antula 2010 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Safim, 2010, intermediate. 33 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-GW Safim 2010 intermediate 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Kilifi, 2012, intermediate. 64 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-KE Kilifi 2012 intermediate 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Kababougou, 2014, gambiae. 40 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-ML-A Kababougou 2014 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Kela, 2004, gambiae. 23 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-ML-B Kela 2004 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Douna, 2004, coluzzii. 20 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-ML-B Douna 2004 coluzzii 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Chikhwawa, 2015, arabiensis. 41 individuals, (arabiensis filter)\n",
      "Converting to .dat AG1000G-MW Chikhwawa 2015 arabiensis 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Furvela, 2004, gambiae. 71 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-MZ Furvela 2004 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Muleba, 2015, arabiensis. 170 individuals, (arabiensis filter)\n",
      "Converting to .dat AG1000G-TZ Muleba 2015 arabiensis 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Moshi, 2012, arabiensis. 40 individuals, (arabiensis filter)\n",
      "Converting to .dat AG1000G-TZ Moshi 2012 arabiensis 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Tarime, 2012, arabiensis. 47 individuals, (arabiensis filter)\n",
      "Converting to .dat AG1000G-TZ Tarime 2012 arabiensis 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Muheza, 2013, gambiae. 43 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-TZ Muheza 2013 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Nagongera, 2012, arabiensis. 194 individuals, (arabiensis filter)\n",
      "Converting to .dat AG1000G-UG Nagongera 2012 arabiensis 3L\n",
      "Writing .batch.txt file for LDNe...\n",
      "\n",
      "Producing LDNe input for Kihihi, 2012, gambiae. 96 individuals, (gambcolu filter)\n",
      "Converting to .dat AG1000G-UG Kihihi 2012 gambiae 3L\n",
      "Writing .batch.txt file for LDNe...\n"
     ]
    }
   ],
   "source": [
    "print(f\"------------------------ LDNe_Ag -------------------------------\\n\")\n",
    "print(f\"Storing metadata in ../data/Phase3.LDNe.tsv\")\n",
    "with open(f'../data/Phase3.LDNe.tsv', 'w') as metafile:\n",
    "    metafile.write(\"sample_set\\tlocation\\tyear\\tspecies\\tchromosome\\n\")\n",
    "\n",
    "Path('../data/Phase3.LDNe.list').touch()    \n",
    "    \n",
    "#filter the gff3 to be coding and regulatory regions\n",
    "df = allel.gff3_to_dataframe(f\"../../data/reference/gff/An.gambiae-PEST-BASEFEATURES_agamP4.12.gff3.gz\")\n",
    "coding_reg_df = df[~df.type.isin(['chromosome', 'three_prime_UTR','five_prime_UTR',\n",
    "                  'mRNA', 'CDS', 'exon'])].drop(columns=['source', 'strand', 'phase', 'score'])\n",
    "coding_reg_df = coding_reg_df[coding_reg_df.seqid == chrom]\n",
    "   \n",
    "\n",
    "for _set in all_sets:\n",
    "  \n",
    "    #subset metadata\n",
    "    metadata = manifest[manifest.sample_set == _set].reset_index(drop=True)\n",
    "\n",
    "    ### loop through combos \n",
    "    for loc in metadata.location.unique():\n",
    "\n",
    "        nmeta = metadata[metadata.location == loc]\n",
    "\n",
    "        for yr in nmeta.year.unique():\n",
    "\n",
    "            nmeta2 = nmeta[nmeta.year == yr]\n",
    "            \n",
    "            #have edited .species_gambiae_coluzzii column to contain arabiensis instead of NA \n",
    "            for sp in nmeta2.species_gambiae_coluzzii.unique():\n",
    "\n",
    "                #if there is less than 15 samples than skip\n",
    "                if (nmeta2.species_gambiae_coluzzii == sp).sum() < 15:\n",
    "                    continue\n",
    "                \n",
    "                if sp == 'arabiensis':  \n",
    "                    print(f\"\\nProducing LDNe input for {loc}, {yr}, {sp}. {nmeta.shape[0]} individuals, (arabiensis filter)\")\n",
    "                    geno, positions = load_arrays_noncoding_and_centromeres(_set, chrom, coding_reg_df, sitefilter='arab')\n",
    "                    #filter to species \n",
    "                    nmeta3 = nmeta[nmeta.species_gambiae_coluzzii == sp]\n",
    "                    flt = np.array(nmeta3.index)\n",
    "                    #filter to correct loc, year, species individuals\n",
    "                    gn = geno.take(flt, axis=1)\n",
    "                    \n",
    "                    convert2dat(gn, positions, n, _set, loc, yr, sp, chrom)\n",
    "              \n",
    "                else:\n",
    "                    print(f\"\\nProducing LDNe input for {loc}, {yr}, {sp}. {nmeta.shape[0]} individuals, (gambcolu filter)\")\n",
    "                    geno, positions = load_arrays_noncoding_and_centromeres(_set, chrom, coding_reg_df, sitefilter='gamb_colu')\n",
    "                    #filter to species \n",
    "                    nmeta3 = nmeta[nmeta.species_gambiae_coluzzii == sp]\n",
    "                    flt = np.array(nmeta3.index)\n",
    "                    #filter to correct loc, year, species individuals\n",
    "                    gn = geno.take(flt, axis=1)\n",
    "\n",
    "                    convert2dat(gn, positions, n, _set, loc, yr, sp, chrom)\n",
    "\n",
    "                #write metadata file for samples that are included\n",
    "                with open('../data/Phase3.LDNe.tsv', 'a') as metafile:\n",
    "                    metafile.write(f'{_set}\\t{loc}\\t{yr}\\t{sp}\\t{chrom}\\n')\n",
    "\n",
    "                with open('../data/Phase3.LDNe.list', 'a') as listfile:\n",
    "                    listfile.write(f'{_set}.{loc}.{yr}.{sp}\\n')\n",
    "                \n",
    "                print(\"Writing .batch.txt file for LDNe...\")\n",
    "                with open(f'../analysis/LDNe/batch/{_set}.{loc}.{yr}.{sp}.{chrom}.batch.txt', 'w') as batch_file:\n",
    "                    batch_file.write(f'1\\t0\\n1\\n0.05\\t-1\\n15\\t0\\t1\\n1\\n0\\n0\\n0\\n0\\nanalysis/LDNe/Ag_LDNe_{_set}.{loc}.{yr}.{sp}.{chrom}.out\\n')\n",
    "                    batch_file.write(f\"data/dat/{_set}.{loc}.{yr}.{sp}.{chrom}.dat\\n\")\n",
    "                    batch_file.write(\"*\")\n",
    "                    batch_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert2dat(gn, positions, n, _set, loc, yr, sp, chrom):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function takes a genotyping array and positions, along with info on the sample_set, location , year, species, chrom,\n",
    "    and converts the genotyping array into a .dat file, suitable for use with LDNe. It randomly downsamples the genotyping array\n",
    "    by n SNPs\n",
    "    \"\"\"\n",
    "    \n",
    "    # MAF 0.05 filter\n",
    "    ac = gn.count_alleles()\n",
    "    freqs = ac.to_frequencies().compute()\n",
    "    ALT1 = freqs[:,1] > 0.05\n",
    "    ALT2 = freqs[:,2] > 0.05\n",
    "    ALT3 = freqs[:,3] > 0.05\n",
    "    maf_flt = np.logical_or(ALT1, ALT2, ALT3)\n",
    "    gn = gn.compress(maf_flt, axis=0)\n",
    "\n",
    "    #take random sample of n SNPs \n",
    "    snp_sample = np.random.choice(gn.shape[0], n, replace=False)\n",
    "    snp_sample.sort()\n",
    "    gnr = gn.take(snp_sample, axis=0)\n",
    "    gnr = np.array(gnr[:])\n",
    "    gnr = gnr.astype(str)\n",
    "\n",
    "    pos = positions[maf_flt]\n",
    "    pos = pos[snp_sample]\n",
    "    prefix = f'{chrom}_'\n",
    "    pos_string = [prefix + p for p in pos.astype(str)]\n",
    "\n",
    "    gnr[gnr == '-1'] = '00' #convert missing alleles \n",
    "    dat = np.empty([gnr.shape[0], gnr.shape[1]])\n",
    "\n",
    "    #join genotypes in same individual \n",
    "    print(f\"Converting to .dat {_set} {loc} {yr} {sp} {chrom}\")\n",
    "    for x in range(gnr.shape[0]):\n",
    "        for y in range(gnr.shape[1]):\n",
    "            dat[x,y] = ''.join(gnr[x,y])\n",
    "\n",
    "    #convert to .dat format genotypes \n",
    "    dat = dat.astype(str)\n",
    "    dat[dat == '0.0'] = '0101'\n",
    "    dat[dat == '1.0'] = '0102'\n",
    "    dat[dat == '2.0'] = '0103'\n",
    "    dat[dat == '3.0'] = '0104'\n",
    "    dat[dat == '10.0'] = '0201'\n",
    "    dat[dat == '11.0'] = '0202'\n",
    "    dat[dat == '12.0'] = '0203'\n",
    "    dat[dat == '13.0'] = '0204'\n",
    "    dat[dat == '20.0'] = '0301'\n",
    "    dat[dat == '21.0'] = '0302'\n",
    "    dat[dat == '22.0'] = '0303'\n",
    "    dat[dat == '23.0'] = '0304'\n",
    "    dat[dat == '30.0'] = '0401'\n",
    "    dat[dat == '31.0'] = '0402'\n",
    "    dat[dat == '32.0'] = '0403'\n",
    "    dat[dat == '33.0'] = '0404'\n",
    "\n",
    "    #stack population name and transposed genotypes \n",
    "    popnames = np.repeat(f\"{_set}{loc}{yr}{sp}{chrom}\", gnr.shape[1])\n",
    "    dat = np.column_stack((popnames, dat.T)) #\n",
    "\n",
    "    #write out .dat file for LDNe \n",
    "    with open(f'../data/dat/{_set}.{loc}.{yr}.{sp}.{chrom}.dat', 'w') as datfile:\n",
    "        datfile.write(f'{gnr.shape[1]}\\t{gnr.shape[0]}\\t4\\t2\\n')\n",
    "        datfile.write(\"\\n\".join(\"\".join(map(str, x)) for x in pos_string)) \n",
    "        datfile.write(\"\\n\")\n",
    "        datfile.write(\"\\n\".join(\"\\t\".join(map(str, x)) for x in dat))\n",
    "        \n",
    "        \n",
    "\n",
    "def load_arrays_noncoding_and_centromeres(local_path, _set, chrom, coding_reg_df, sitefilter='gamb_colu'):\n",
    "    \n",
    "    \"\"\"\n",
    "    This function reads and filters a genotyping array to the noncoding, noncentromeric regions, and applys a filter depending on \n",
    "    whether the samples are arabiensis (arab) or gambiae/coluzzii (gamb_colu)\n",
    "    \"\"\"\n",
    "    \n",
    "    Ag_array = zarr.open_array(f\"{local_path}/snp_genotypes/all/{_set}/{chrom}/calldata/GT/\", mode = 'r')\n",
    "    filters = zarr.open(f\"{local_path}/site_filters/dt_20200416/{sitefilter}/{chrom}/variants/filter_pass\", mode=\"r\")\n",
    "    positions = zarr.open_array(f\"{local_path}/snp_genotypes/all/sites/{chrom}/variants/POS/\", mode='r')\n",
    "    positions = positions[:][filters[:]]    \n",
    "    geno = allel.GenotypeDaskArray(Ag_array)\n",
    "    geno = geno[filters[:]]\n",
    "    \n",
    "    if chrom == '2L':\n",
    "        centromere = (positions > 3000000)\n",
    "    elif chrom == '2R':\n",
    "        centromere = (positions < 57000000)\n",
    "    elif chrom == '3L':\n",
    "        centromere = (positions > 2000000)\n",
    "    elif chrom == '3R':\n",
    "        centromere = (positions < 50000000)\n",
    "    elif chrom == 'X':\n",
    "        centromere = (positions < 21000000) \n",
    "        \n",
    "    positions = allel.SortedIndex(positions[centromere])\n",
    "    #get boolean array for positions that are coding - allel.locate_ranges so fast!\n",
    "    coding = positions.locate_ranges(coding_reg_df.start, coding_reg_df.end, strict=False)\n",
    "    #compress to get noncoding SNPs and remove centromeric regions of low recombination\n",
    "    #get non-centromeric regions. currently chosen by eye based on ag1000g phase1 paper fig1.\n",
    "  \n",
    "    #TODO currently using all non-coding regions, alternate option may be to take SNPs x distance from coding regions \n",
    "    geno = geno.compress(centromere, axis=0)\n",
    "    geno = geno.compress(~coding, axis=0) #we want noncoding regions so '~' to get inverse of boolean\n",
    "    positions = positions[~coding]\n",
    "    \n",
    "    return(geno, positions)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
